subsys/audio_module/audio_module.h

```diff
@@ -0,0 +1,463 @@
+				 struct audio_module_configuration *configuration);
+
+	/**
+	 * @brief Start a module.
+	 *
+	 * @param handle[in/out]  The handle for the module to start
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*start)(struct audio_module_handle_private *handle);
+
+	/**
+	 * @brief Stop a module.
+	 *
+	 * @param handle[in/out]  The handle for the module to stop
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*stop)(struct audio_module_handle_private *handle);
+
+	/**
+	 * @brief The core data processing function for the module. Can be either an
+	 *		  input, output or processor module type.
+	 *
+	 * @param handle[in/out]  The handle to the module instance
+	 * @param block_rx[in]    Pointer to the input audio block or NULL for an input module
+	 * @param block_tx[out]   Pointer to the output audio block or NULL for an output module
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*data_process)(struct audio_module_handle_private *handle, struct audio_data *block_rx,
+			    struct audio_data *block_tx);
+};
+
+/**
+ * @brief A module's minimum description.
+ */
+struct audio_module_description {
+	/* The module base name */
+	char *name;
+
+	/* The module type */
+	enum audio_module_type type;
+
+	/* A pointer to the functions in the module */
+	const struct audio_module_functions *functions;
+};
+
+/**
+ * @brief Module's thread configuration structure.
+ */
+struct audio_module_thread_configuration {
+	/* Thread stack */
+	k_thread_stack_t *stack;
+
+	/* Thread stack size */
+	size_t stack_size;
+
+	/* Thread priority */
+	int priority;
+
+	/* A pointer to a module's data receiver FIFO, can be NULL */
+	struct data_fifo *msg_rx;
+
+	/* A pointer to a module's data transmitter FIFO, can be NULL */
+	struct data_fifo *msg_tx;
+
+	/* A pointer to the data buffer slab, can be NULL */
+	struct k_mem_slab *data_slab;
+
+	/* Size of each memory block in bytes that will be
+	 * taken from the data buffer slab or 0
+	 */
+	size_t data_size;
+};
+
+/**
+ * @brief Module's generic set-up structure.
+ */
+struct audio_module_parameters {
+	/* The module's private description */
+	struct audio_module_description *description;
+
+	/* The module's thread setting */
+	struct audio_module_thread_configuration thread;
+};
+
+/**
+ * @brief Private module handle.
+ */
+struct audio_module_handle {
+	/* The unique name of this module instance */
+	char name[CONFIG_AUDIO_MODULE_NAME_SIZE];
+
+	/* The module's description */
+	struct audio_module_description *description;
+
+	/* Current state of the module */
+	enum audio_module_state state;
+
+	/* Previous state of the module */
+	enum audio_module_state previous_state;
+
+	/* Thread ID */
+	k_tid_t thread_id;
+
+	/* Thread data */
+	struct k_thread thread_data;
+
+	/* Flag to indicate if the module should send data block to its TX FIFO */
+	bool data_tx;
+
+	/* List node (next pointer) */
+	sys_snode_t node;
```

Comment:
What is the difference between this one and the one on line 261?

Reply:
The node gives the "next" pointer and the hdl_dest_list is the head and tail of the list. This allows me to scan the list to perform an operation on each element, the handle of the module to send the data to in this case.
i.e. SYS_SLIST_FOR_EACH_CONTAINER(&handle->hdl_dest_list, handle_to, node) {......}